### PR TITLE
Async mq

### LINKED ISSUE #10: Create eventbus plugin

### PR DESCRIPTION
**Pull request by Kargathia**
Resolves #10 

Changelog:

* Added `events.py` for AMQP subscribing/publishing
* Moved responsibility for prefix from gateway to service
* Upped version number to `0.3.1`
* Renamed `simulator.init_app(app)` to `simulator.setup(app)` to keep it in line with aiohttp convention

Shared behavior:

Calling `events.setup(app)` before app started will add a `EventListener` and a `EventPublisher` to app features. 
If listeners/publishers are initialized with a pre-run app as argument, they will hook themselves up to the app `on_startup` and `on_cleanup` events.
clients initializing listeners/publishers after app start will need to call `start(loop)` and `close()` functions explicitly.
If connection is lost, listeners/publishers will automatically keep trying to reconnect. All connections and subscriptions are restored automatically on reconnect.

EventListener behavior:

Subscriptions can be made at any time using `EventListener.subscribe()`. They will be declared on the remote amqp server whenever the listener is connected. 
Message interest can be specified by setting exchange name, and routing key.

For `direct` and `fanout` exchanges, messages must match routing key exactly. For `topic` exchanges (the default), routing keys can be nested values using `.`, and `*` and `#` wildcards are allowed.

For more information on this, see the [routing tutorial](https://www.rabbitmq.com/tutorials/tutorial-four-python.html), and the [topic tutorial](https://www.rabbitmq.com/tutorials/tutorial-five-python.html).

The listener will attempt to parse incoming messages as JSON.

EventPublisher behavior:

Whenever it is connected, `EventPublisher` can publish messages to the server. For each message, exchange and routing key can be specified.
If the publisher is not currently connected, it will raise a `ConnectionRefusedError`.

Published messages will be serialized as JSON.

### CODE REVIEW

File: new/brewblox_service/announcer.py [hunk 1]
```diff
@@ -34,22 +22,22 @@
         'name': name,
         'active': True,
         'proxy': {
-            # Strips 'listen_path'
-            'strip_path': True,
+            # listen path is part of actual endpoint
```
Comment (elcojacobs) [inline new/brewblox_service/announcer.py]: Is this reworded correctly? path on this service is equal to path on gateway. The gateway only proxies on host:port level and does not rewrite url/endpoint.

File: new/brewblox_service/events.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            auto_delete=True)
+
+        # Push it over the line
+        await exchange.publish(
+            aio_pika.Message(data),
+            routing_key=routing
+        )
+
+
+@routes.post('/publish')
```
Comment (elcojacobs) [inline new/brewblox_service/events.py]: This is temporary for demonstration, right? Should move out of library and into specific plugins?

File: new/brewblox_service/service.py [hunk 3]
```diff
@@ -117,8 +122,29 @@ def furnish(app: Type[web.Application]):
...
+            cors.add(route)
+            LOGGER.debug(f'Enabled CORS for {route.resource}')
+
     # Configure swagger settings
+    # We set prefix explicitly here
     aiohttp_swagger.setup_swagger(app,
+                                  swagger_url=prefix + '/api/doc',
                                   description='',
                                   title='Brewblox Service',
                                   api_version='0.3.0',
```
Comment (elcojacobs) [inline new/brewblox_service/service.py]: Do we have a plan for api versioning per service? Should services override this? Or do we have a version for the service itself and for the import service library?

File: new/brewblox_service/announcer.py (hunk 1 above)
Comment (Kargathia) [reply new/brewblox_service/announcer.py]: Yes. Example:
* Gateway url = `http://gateway:1234`
* Controller url = `http://controlhost:5678`
* Controller service name = `my_controller`

`http://gateway:1234/my_controller/api/doc` will now proxy to `http://controlhost:5678/my_controller/api/doc`. 

The gateway `listen_path` is `my_controller`. This should also be part of the proxied-to url.

File: new/brewblox_service/events.py (hunk 2 above)
Comment (Kargathia) [reply new/brewblox_service/events.py]: Kept as debugging tool for services.

File: new/brewblox_service/service.py (hunk 3 above)
Comment (Kargathia) [reply new/brewblox_service/service.py]: Swagger api version should equal service implementation version. All our versioning - and linking to automatic deployment is todo.

File: new/brewblox_service/announcer.py (hunk 1 above)
Comment (elcojacobs) [reply new/brewblox_service/announcer.py]: example makes it very clear, can be copied directly to code comments

**Pull request merged**